src/ripple/app/misc/impl/TxQ.cpp

```diff
@@ -107,16 +108,31 @@ TxQ::FeeMetrics::update(Application& app,
         // so clamp down on limits.
         txnsExpected_ = boost::algorithm::clamp(feeLevels.size(),
             minimumTxnCount_, targetTxnCount_);
+        recentTxnCounts_.clear();
     }
     else if (feeLevels.size() > txnsExpected_ ||
         feeLevels.size() > targetTxnCount_)
     {
+        recentTxnCounts_.emplace_back(feeLevels.size());
+        while (recentTxnCounts_.size() > setup.ledgersInQueue)
+            recentTxnCounts_.pop_front();
+        auto const iter = std::max_element(recentTxnCounts_.begin(),
+            recentTxnCounts_.end());
+        BOOST_ASSERT(iter != recentTxnCounts_.end());
+        auto const next = [&]
+        {
+            if (*iter >= txnsExpected_)
+                return *iter;
+            // 90% backoff to decrease
+            // down to the actual max.
+            return (txnsExpected_ * 9 + *iter) / 10;
+        }();
         // Ledgers are processing in a timely manner,
         // so keep the limit high, but don't let it
         // grow without bound.
         txnsExpected_ = maximumTxnCount_ ?
-            std::min(feeLevels.size(), *maximumTxnCount_) :
-            feeLevels.size();
+            std::min(next, *maximumTxnCount_) :
+            next;
```

Comment:
This seems to be overly complex with no good reason. Consider:

`txnsExpected_ = std::min(next, maximumTxnCount_.value_or(next));`

Reply:
Much better.